### PR TITLE
Dasd 8310 add building blocks

### PR DESCRIPTION
**Pull request by NickGraham101**

### CODE REVIEW

File: templates/public-ip-address.json [hunk 1]
```diff
@@ -49,6 +49,10 @@
     "PublicIPAddress": {
       "type": "string",
       "value": "[reference(resourceId('Microsoft.Network/publicIPAddresses',parameters('publicIPAddressName')),providers('Microsoft.Network', 'publicIPAddresses').apiVersions[0]).ipAddress]"
+    },
+    "PublicIpAddressId": {
```
Comment (jack-education) [inline templates/public-ip-address.json:53]:
```suggestion
    "PublicIpAddressResourceId": {
```

File: templates/app-gateway-v2.json [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "appGatewayName": {
+            "type": "string",
+            "metadata": {
+                "description": "Name of the application gateway resource"
+            }
+        },
+        "subnetRef": {
```
Comment (jack-education) [inline templates/app-gateway-v2.json:11]:
```suggestion
        "subnetResourceId": {
```

File: templates/app-gateway-v2.json [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    },
+    "variables": {
+        "logDiagnosticEnabled": "[or(not(empty(parameters('logStorageAccountId'))),not(empty(parameters('logWorkspaceId'))))]",
+        "logRetentionEnabled": "[if(equals(parameters('logRetention'),0), json('false'), json('true'))]",
+        "tier": {
+            "Standard_Small": "Standard",
+            "Standard_Medium": "Standard",
+            "Standard_Large": "Standard",
+            "WAF_Medium": "WAF",
+            "WAF_Large": "WAF",
```
Comment (jack-education) [inline templates/app-gateway-v2.json:134]:
```suggestion
```
Can these be removed so the template is specific to App Gateway v2?

File: templates/managed-identity.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        {
+            "name": "[parameters('userAssignedIdentityName')]",
+            "type": "Microsoft.ManagedIdentity/userAssignedIdentities",
+            "apiVersion": "2018-11-30",
+            "location": "[resourceGroup().location]"
+        }
+    ],
+    "outputs": {
+    }
+}
```
Comment (jack-education) [inline templates/managed-identity.json:21]: We have https://github.com/SkillsFundingAgency/das-platform-building-blocks/blob/master/templates/user-assigned-identity.json

File: templates/public-ip-address.json (hunk 1 above)
Comment (NickGraham101) [reply templates/public-ip-address.json]: Updated

File: templates/app-gateway-v2.json (hunk 2 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json:11]: Replaced with 3 variables

File: templates/app-gateway-v2.json (hunk 3 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json:134]: Removed

File: templates/managed-identity.json (hunk 4 above)
Comment (NickGraham101) [reply templates/managed-identity.json:21]: deleted template

File: templates/app-gateway-v2.json [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+                "description": "Set the private IP address for application gateway (public IP address only generated if empty)"
+            }
+        },
+        "publicIpAddressId": {
+            "type": "string",
+            "defaultValue": "",
+            "metadata": {
+                "description": "Reference ID for a public IP address"
+            }
+        },
```
Comment (jack-education) [inline templates/app-gateway-v2.json:92]:
```suggestion
        "publicIpAddressResourceId": {
            "type": "string",
            "defaultValue": "",
            "metadata": {
                "description": "Resource ID for a public IP address"
            }
        },
```

File: templates/app-gateway-v2.json [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            "Standard_v2": "Standard_v2",
+            "WAF_v2": "WAF_v2"
+        },
+        "frontendIp": {
+            "public": [
+                {
+                    "name": "appGatewayPublicFrontendIp",
+                    "properties": {
+                        "PublicIPAddress": {
+                            "id": "[parameters('publicIpAddressId')]"
```
Comment (jack-education) [inline templates/app-gateway-v2.json:142]:
```suggestion
                            "id": "[parameters('publicIpAddressResourceId')]"
```

File: templates/app-gateway-v2.json [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+                    {
+                        "name": "appGatewayIpConfig",
+                        "properties": {
+                            "subnet": {
+                                "id": "[variables('subnetResourceId')]"
+                            }
+                        }
+                    }
+                ],
+                "frontendIPConfigurations": "[if(and(greater(length(parameters('privateIpAddress')),0),greater(length(parameters('publicIpAddressId')),0)), union(variables('frontendIp').public,variables('frontendIp').private), if(greater(length(parameters('privateIpAddress')),0), variables('frontendIp').private, if(greater(length(parameters('publicIpAddressId')),0), variables('frontendIp').public, variables('blankArray'))))]",
```
Comment (jack-education) [inline templates/app-gateway-v2.json:281]:
```suggestion
                "frontendIPConfigurations": "[if(and(greater(length(parameters('privateIpAddress')),0),greater(length(parameters('publicIpAddressResourceId')),0)), union(variables('frontendIp').public,variables('frontendIp').private), if(greater(length(parameters('privateIpAddress')),0), variables('frontendIp').private, if(greater(length(parameters('publicIpAddressResourceId')),0), variables('frontendIp').public, variables('blankArray'))))]",
```

File: templates/app-gateway-v2.json (hunk 6 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json]: updated

File: templates/app-gateway-v2.json (hunk 7 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json]: updated

File: templates/app-gateway-v2.json (hunk 5 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json]: updated

File: templates/app-gateway-v2.json [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+            "defaultValue": "",
+            "metadata": {
+                "description": "Set the private IP address for application gateway (public IP address only generated if empty)"
+            }
+        },
+        "publicIpAddressResourceId": {
+            "type": "string",
+            "defaultValue": "",
+            "metadata": {
+                "description": "Reference ID for a public IP address"
```
Comment (jack-education) [inline templates/app-gateway-v2.json:90]:
```suggestion
                "description": "Resource ID for a public IP address"
```

File: templates/app-gateway-v2.json [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+                "description": "The name of the user assigned identity."
+            }
+        },
+        "logStorageAccountId": {
+            "type": "string",
+            "defaultValue": "",
+            "metadata": {
+                "description": "Storage account to archive logs to (leave blank to disable)"
+            }
+        },
```
Comment (jack-education) [inline templates/app-gateway-v2.json:119]: Can this final resourceId be a variable using storage account name / resource group parameters

File: templates/app-gateway-v2.json [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+            "properties": {
+                "sku": {
+                    "name": "[parameters('appGatewayTier')]",
+                    "tier": "[variables('tier')[parameters('appGatewayTier')]]",
+                    "capacity": "[parameters('capacity')]"
+                },
+                "sslPolicy": {
+                    "policyType": "Predefined",
+                    "policyName": "AppGwSslPolicy20170401S"
+                },
```
Comment (jack-education) [inline templates/app-gateway-v2.json:275]: Can this be a custom sslPolicy defined in a variable to match das-firewalls.?

File: templates/app-gateway-v2.json (hunk 9 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json:119]: I've removed this as I don't anticipate it being used but I've refactored the OMS related params to be consistent with the established pattern

File: templates/app-gateway-v2.json [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+            }
+        }
+    },
+    "variables": {
+        "logDiagnosticEnabled": "[or(not(empty(parameters('logStorageAccountId'))),not(empty(parameters('logWorkspaceId'))))]",
+        "logRetentionEnabled": "[if(equals(parameters('logRetention'),0), json('false'), json('true'))]",
+        "tier": {
+            "Standard_v2": "Standard_v2",
+            "WAF_v2": "WAF_v2"
+        },
```
Comment (jack-education) [inline templates/app-gateway-v2.json:141]: Is this variable needed? Can the parameter be used directly?

Comment (NickGraham101) [reply templates/app-gateway-v2.json]: Not needed, removed

File: templates/app-gateway-v2.json (hunk 10 above)
Comment (NickGraham101) [reply templates/app-gateway-v2.json]: Implemented as param consuming az devops variable

**Review by jack-education — APPROVED** [review]

**Pull request merged**